Pytest using pyproject.toml.

Paebbels opened the pull request:
# New Features
* Allow configuration of the unit test directory (default: `tests/unit`).
* Allow configuration of a `pyproject.toml` or `.coveragerc` file.
* Extract values from `pyproject.toml` or `.coveragerc`.

# Changes
* Jobs deriving from template job `CoverageCollection` need to specify their `.coveragerc` content and `pytest.ini` content in a `pyproject.toml` file or provide the job parameter `coverage_config` with the path to the `.coveragerc` file.

# Bug Fixes
*None*

-------------
* Closes #2.

Paebbels commented on .github/workflows/Parameters.yml:
```diff
@@ -76,7 +76,7 @@ jobs:
 
           data = {
-             '3.6': { 'icon': '🔴', 'until': '23.12.2021' },
+             '3.6': { 'icon': '🔴', 'until': '23.12.2021' },   # Black circle ⚫ for EOL versions.
```
I propose to set Python 3.6 to a black icon in the next days for 1 month and then deprecate Python 3.6. Colors are then moved by one position.

umarcor replied:
Agree. However, I think that should be an explicit PR.

Paebbels replied:
The change will be a separate PR. I just added the colored circle to have it in the file.

Paebbels commented:
@umarcor the branch `toml` is used and tested here: https://github.com/pyTooling/pyTooling/actions/runs/1608056278

This repos with `dev` branch has only a single configuration file as `pyproject.toml`. No `pytest.ini` and no `.coveragerc` anymore.

umarcor commented on .github/workflows/CoverageCollection.yml:
```diff
@@ -54,6 +64,26 @@ jobs:
+      - name: 🔁 Extract configurations from pyproject.toml
+        id: getVariables
+        run: |
+          function ReadToml() {
```
> Convert to either JavaScript (or Python?) or encapsulate in an Action?

Just add `shell: python` in the step (maybe after `id` and before `run`) and use Python straightaway. See https://github.com/pyTooling/Actions/blob/main/.github/workflows/Parameters.yml#L59-L62.

> It should only run if coveragerc is not given, because then it defaults to pyproject.toml.

Use `if: input.coverage_config == ''`. See https://github.com/pyTooling/Actions/blob/main/.github/workflows/UnitTesting.yml#L74.

> At best, also for coveragerc, we should read the values from there.

Which values do you want/need to read from coveragerc?

Paebbels replied:
coveragerc contains the same content:
* output directory of the HTML report.
* output path to the XML report file.

With having them extracted, a user can have a local setup that fits his needs, but CI does run independent and can collect generated outputs without needing `input` variables to double encode these paths.

Paebbels replied:
I modified the code on the branch to use Python in the Python "shell". Now I need a dependency installed (`tomli`).  
See https://github.com/pyTooling/pyTooling/runs/4607370478?check_suite_focus=true#step:3:37

The Python shell works in the job even without Python installed via the Python setup action. Any idea how to install a requirement, because both Pythons seam to be isolated.

umarcor replied:
Use one step before the python script:

```
- shell: bash
  run: pip install tomli
```

By default, Python is available on the ubuntu-latest environment. Using the setup-python Action is not necessary. You need to use it only in case you want to explicitly customise the version of Python to be used.

In this case, since setup-python is used later, you might want to reorder the steps:

- First, setup-python.
- Second, pip.
- Third the python script.

Paebbels replied:
Fixed in fa10ed07.

umarcor commented on .github/workflows/CoverageCollection.yml:
```diff
@@ -67,30 +97,31 @@ jobs:
         run: |
           coverage html
-          rm htmlcov/.gitignore
+          rm ${{ steps.getVariables.outputs.coverage_report_html_directory }}/.gitignore
```
I think that the content is flatenned in the zipfile. Hence, the path used here should have no effect when downloading the artifact.

Paebbels replied:
ok. In GitLab the path is preserved, which ensures the artifact is recreated in another job at the same location :)
(which I very like - that's the purpose of artifacts - job-to-job data transfer ...).

umarcor replied:
In GitHub, job-to-job data transfer is just one of the usages of artifacts. Hence, there is the `path` field when downloading the artifact, in case you want to customise the location. That is actually useful in several cases, because it avoids an step to move the content (either before or after). Typical case is producing multiple HTML artifacts and then downloading them in a given hierarchy before pushing to GitHub Pages: https://github.com/pyTooling/Actions/blob/main/.github/workflows/PublishToGitHubPages.yml#L53-L71

umarcor commented on .github/workflows/CoverageCollection.yml:
```diff
@@ -67,30 +97,31 @@ jobs:
         with:
           name: ${{ inputs.artifact }}
-          path: htmlcov
+          path: ${{ steps.getVariables.outputs.coverage_report_html_directory }}
```
This won't work if step `getVariables` is executed conditionally. Therefore, it needs to be executed always, and the condition can be used to decide the content of the outputs. The same applies to the other output used twice below.

Paebbels replied:
Conditional execution isn't planned anymore. The Step will always run and extract the information from either source.

umarcor commented on .idea/Actions.iml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please, add this in one specific commit, and mention it in the changelog.

Paebbels replied:
Strange that this file was added ...

Paebbels replied:
I removed that file.

umarcor commented:
> @umarcor the branch `toml` is used and tested here: https://github.com/pyTooling/pyTooling/actions/runs/1608056278
> 
> This repos with `dev` branch has only a single configuration file as `pyproject.toml`. No `pytest.ini` and no `.coveragerc` anymore.

I think it should be tested on three repos at the same time:

- With `pyproject.toml` only.
- With `.coverager` only.
- None of them.

Paebbels commented:
@umarcor The test pipeline was now successful: https://github.com/pyTooling/pyTooling/actions/runs/1619099974

umarcor commented:
@Paebbels there seems to be some issue with the location of the artifacts/reports: https://github.com/pyTooling/pyTooling/runs/4626643488?check_suite_focus=true. Moreover, I think it is important to test other use cases: https://github.com/pyTooling/Actions/pull/29#issuecomment-1000796852.

Paebbels commented:
It's a race condition. See:
![image](https://user-images.githubusercontent.com/956109/147354759-3153d925-0f31-458d-b369-56ff9632088b.png)

The artifact cleanup has a missing `needs` for the XML test result publishing.

Paebbels commented:
Yes we can add a test project here (or multiple - not sure if it fits into one repo ...?).

Anyhow, The current change is tested. So I would delay this PR.
I have currently a chain of 3 more PR waiting for this one :).
(I know it's not a good answer - but we shouldn't enforce better quality for this repo in this PR, when there is currently no test at all.)

umarcor commented:
> It's a race condition.
> The artifact cleanup has a missing `needs` for the XML test result publishing.

Example fixed in https://github.com/pyTooling/Actions/commit/66c7b4b619465aea9f3a6337c3db2244ab608fd6.

umarcor commented:
> Yes we can add a test project here (or multiple - not sure if it fits into one repo ...?).
> 
> Anyhow, The current change is tested. So I would delay this PR. I have currently a chain of 3 more PR waiting for this one :).

It's not about testing in this repo. That is desirable, but it's an smoke-test only.

This PR (and others) need to be tested on other repos, by changing the version of the reusable workflows to `@toml`.
That is because we use `uses: actions/checkout@v2` without arguments, so it clones the repo it is being executed on.
Moreover, this PR is extending the supported use cases. Hence, a single repo is not enough because we need one with each of the expected configuration files (or without them).

We might merge this to `dev` without testing, but we should not release and bump `r0`. Therefore, it will not unblock other PRs in different repos, unless those use the development version.

> (I know it's not a good answer - but we shouldn't enforce better quality for this repo in this PR, when there is currently no test at all.)

All previous PRs/enhancements except #16 were tested on multiple repos. Precisely, we had to release v0.2.1 for reverting #16 because it had gone into the release untested. Therefore, even though I understand it will delay other PRs, testing is important.

Paebbels commented:
in pyTooling, `@toml` is used to test the changes.

umarcor commented:
> in pyTooling, `@toml` is used to test the changes.

We need to test also in a repo with `.coveragerc` (no `pyproject.toml`). Do we have any? Or did we update all of them already?

Paebbels commented:
Only pyTooling was changed. The others still use coveragerc and pytest.ini

umarcor commented:
See:

- https://github.com/edaa-org/pySVModel/runs/4627411265?check_suite_focus=true
- https://github.com/edaa-org/pySVModel/commit/734675fe8f36341cc4a44835669c82d473128a95

umarcor commented on .github/workflows/CoverageCollection.yml:
```diff
@@ -35,6 +35,16 @@ on:
+      coverage_config:
+        description: 'Path to the .coveragerc file. Use pyproject.toml if empty.'
+        required: false
+        default: ''
```
It seems that this is a breaking change because it requires existing workflows to explicitly declare `coverage_config: tests/.coveragerc`?

umarcor replied:
I tried declaring it, but it is not handling it:

- https://github.com/edaa-org/pySVModel/runs/4627440518?check_suite_focus=true
- https://github.com/edaa-org/pySVModel/commit/5b8f42a661ef7fae572cba1b5427c63b54b57ff4

umarcor replied:
So, the `.coveragerc` is not a valid TOML file. It should be:

```toml
[run]
branch = true

[report]
skip_covered = true
skip_empty = true

[html]
directory = '.cov'

[xml]
output = 'coverage.xml'
```

umarcor replied:
With the fixed file format, it is read, but there is an issue with the variable name: https://github.com/edaa-org/pySVModel/runs/4627481390?check_suite_focus=true. It should be `coverageRCFile` instead of `pyProjectSettings`?

umarcor replied:
After fixing the variable name and passing arg `-d` to `coverage html`, pySVModel works with this branch: https://github.com/edaa-org/pySVModel/runs/4627543532?check_suite_focus=true.

umarcor replied:
The last use case would be testing this on a repo without `pyproject.toml` and without `.coveragerc`. Do we have any repo of that kind?

Paebbels replied:
Yes it's a little bit breaking. But I think it would be good to migrate towards pyproject.toml and reduce config files with maybe redundant information.

Paebbels replied:
All out projects either have that file, so we just need to handover the parameter or we transition immediately to pyproject.toml in such a case (preferred).

umarcor replied:
Do you expect to force users to have coverage configuration data (either in `pyproject.toml` or in `.coveragerc`)?

Paebbels replied:
Hmmm good question. At least for me it makes almost no sense without them.
But users could have no entries and use defaults ...

umarcor replied:
After pushing some additional fixes, this is tested without coverage configuration files as well:

- https://github.com/edaa-org/pySVModel/runs/4627738759?check_suite_focus=true
- https://github.com/edaa-org/pySVModel/commit/008bdf63ab4d2ea73e3bc4605a7ca7e805678b54

umarcor commented:
I changed this to a draft until we:

- Update the `Changes` section of the first message above.
- Check whether `ExamplePipeline.yml` needs to be updated.
- Check whether `README.md` needs to be updated.

Paebbels commented on .github/workflows/CoverageCollection.yml:
```diff
@@ -62,35 +72,76 @@ jobs:
         run: |
-          coverage html
-          rm htmlcov/.gitignore
+          coverage html -d ${{ steps.getVariables.outputs.coverage_report_html_directory }}
```
Why do you specify the directory with `-d`?
This directory is taken from `pyproject.toml` oder `coveragerc` file.

umarcor replied:
With `.coveragerc`, the report is written in `htmlcov`: https://github.com/edaa-org/pySVModel/runs/4633523272?check_suite_focus=true#step:8:8. It seems that the `html` field of the config file is not taken into account.

umarcor replied:
We decided to support the three use cases for now, but we will recommend migrating to `pyproject.toml`. Other cases might be deprecated and removed in the future.

Paebbels commented:
> I changed this to a draft until we:
> 
> * Update the `Changes` section of the first message above.
> * Check whether `ExamplePipeline.yml` needs to be updated.
> * Check whether `README.md` needs to be updated.

Updated `Changes` section in this PR and also adjusted README in 78b22519.
